### PR TITLE
feat(dashboard): workflow editor error handling + sentry

### PR DESCRIPTION
**Pull request by LetItRock**
### What changed? Why was the change needed?
Dashboard Workflow Editor error handling plus Sentry configuration.

### Screenshots
Screenshot 2024-10-26 at 00 38 41

Screenshot 2024-10-25 at 13 27 53

https://github.com/user-attachments/assets/468a7e15-7b40-4277-b67f-dac193e6b895

https://github.com/user-attachments/assets/deb474a3-31c5-4539-81c3-9062dce358ae

### CODE REVIEW

File: apps/dashboard/legacy-handler.js [hunk 1]
```diff
@@ -9,6 +9,7 @@ export const legacyHandler = (__dirname, req, res, next) => {
     if (fileExists) {
       // Serve the legacy index.html file
       res.setHeader('Content-Type', 'text/html');
+      res.setHeader('Document-Policy', 'js-profiling');
```
Comment (LetItRock) [inline apps/dashboard/legacy-handler.js:12]: Sentry profiling header added in a couple of places, [ref](https://docs.sentry.io/platforms/javascript/guides/react/profiling/)

File: apps/dashboard/package.json [hunk 2]
```diff
@@ -39,6 +39,7 @@
     "@radix-ui/react-tabs": "^1.1.1",
     "@radix-ui/react-tooltip": "^1.1.3",
     "@segment/analytics-next": "^1.73.0",
+    "@sentry/react": "^8.35.0",
```
Comment (LetItRock) [inline apps/dashboard/package.json:42]: sentry react package

File: apps/dashboard/package.json [hunk 3]
```diff
@@ -66,6 +67,7 @@
     "@clerk/types": "^4.6.1",
     "@eslint/js": "^9.9.0",
     "@playwright/test": "^1.44.0",
+    "@sentry/vite-plugin": "^2.22.6",
```
Comment (LetItRock) [inline apps/dashboard/package.json:70]: this is the plugin used during the app build time; it is mainly used to provide the file paths and [components names](https://docs.sentry.io/platforms/javascript/guides/react/features/component-names/)

File: apps/dashboard/src/components/primitives/popover.tsx [hunk 4]
```diff
@@ -30,6 +32,20 @@ const PopoverContent = React.forwardRef<
...
+  React.ComponentPropsWithoutRef<typeof PopoverPrimitive.Arrow> & { portal?: boolean }
+>(({ className, ...props }, ref) => {
+  return (
+    <PopoverPrimitive.Arrow
+      ref={ref}
+      className={cn('arrow fill-background [filter:drop-shadow(0_0_4px_hsl(var(--neutral-alpha-600)))]', className)}
+      {...props}
+    />
+  );
+});
```
Comment (LetItRock) [inline apps/dashboard/src/components/primitives/popover.tsx:47]: added the arrow component to the popover

File: apps/dashboard/src/components/workflow-editor/base-node.tsx [hunk 5]
```diff
@@ -63,11 +65,34 @@ export const NodeBody = ({ children }: { children: ReactNode }) => {
...
+      </PopoverTrigger>
+      <PopoverPortal>
+        <PopoverContent className="flex w-max max-w-[200px] rounded-xl p-2" side="right">
+          <PopoverArrow />
+          <span className="text-destructive text-xs font-normal">{children}</span>
+        </PopoverContent>
+      </PopoverPortal>
+    </Popover>
+  );
+};
```
Comment (LetItRock) [inline apps/dashboard/src/components/workflow-editor/base-node.tsx:89]: this is the error icon that is shown on the workflow step node

File: apps/dashboard/src/components/workflow-editor/workflow-canvas.tsx [hunk 6]
```diff
@@ -60,13 +60,17 @@ const mapStepToNode = (
     content = `Wait to send ~ 30 minutes`;
   }
 
+  const fieldIssues = Object.values({ ...step.issues?.body, ...step.issues?.control })[0];
+  const error = fieldIssues?.[0]?.message;
```
Comment (LetItRock) [inline apps/dashboard/src/components/workflow-editor/workflow-canvas.tsx:64]: this is the interface we have agreed on with BE; looking for the first issue

File: apps/dashboard/src/components/workflow-editor/workflow-editor-provider.tsx [hunk 7]
```diff
@@ -65,6 +66,11 @@ export const WorkflowEditorProvider = ({ children }: { children: ReactNode }) =>
   const { updateWorkflow } = useUpdateWorkflow({
     onSuccess: (data) => {
       reset({ ...data, steps: data.steps.map((step) => ({ ...step })) });
+
+      if (data.issues) {
+        handleValidationIssues({ fields: form.getValues(), issues: data.issues as any, setError });
```
Comment (LetItRock) [inline apps/dashboard/src/components/workflow-editor/workflow-editor-provider.tsx:72]: the `data.issues` is not properly typed on the BE, that's why doing `any` for now

Comment (LetItRock) [reply apps/dashboard/src/components/workflow-editor/workflow-editor-provider.tsx:72]: this utility function sets errors in the form by mapping errors from BE

File: apps/dashboard/src/hooks/use-monitoring.ts [hunk 8]
```diff
@@ -0,0 +1,31 @@
+import { useEffect } from 'react';
+import { setUser as sentrySetUser } from '@sentry/react';
+import { useAuth } from '@/context/auth/hooks';
+
+export function useMonitoring() {
```
Comment (LetItRock) [inline apps/dashboard/src/hooks/use-monitoring.ts:5]: copied from the web app and removed unused code

File: apps/dashboard/src/routes/root.tsx [hunk 9]
```diff
@@ -1,24 +1,44 @@
...
-        <SegmentProvider>
-          <AuthProvider>
-            <HelmetProvider>
-              <Outlet />
-            </HelmetProvider>
-          </AuthProvider>
-        </SegmentProvider>
-      </ClerkProvider>
-    </QueryClientProvider>
+    <ErrorBoundary
```
Comment (LetItRock) [inline apps/dashboard/src/routes/root.tsx:13]: sentry error boundary, we can improve the designs if we want later

File: apps/dashboard/src/utils/handleValidationIssues.ts [hunk 10]
```diff
@@ -14,14 +14,16 @@ type HandleValidationIssuesProps<T extends FieldValues> = {
...
 
   (Object.keys(issues) as Array<keyof typeof issues>).map((issueKey) => {
-    if (issueKey in fields) {
-      setError(issueKey as any, { message: issues[issueKey][0]?.message || 'Unknown error' });
+    const key = issueKey as FieldPath<T>;
+    if (fields[key]) {
+      setError(key, { message: issues[issueKey][0]?.message || 'Unknown error' });
     } else {
-      console.log(`Issue for ${issueKey} found and does not correspond to a field`);
+      console.error(`Issue for ${issueKey} found and does not correspond to a field`);
```
Comment (LetItRock) [inline apps/dashboard/src/utils/handleValidationIssues.ts:26]: sentry is set up to capture all console errors ;)

File: apps/dashboard/src/utils/sentry.ts [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+      integrations: [
+        // See docs for support of different versions of variation of react router
+        // https://docs.sentry.io/platforms/javascript/guides/react/configuration/integrations/react-router/
+        Sentry.reactRouterV6BrowserTracingIntegration({
+          useEffect,
+          useLocation,
+          useNavigationType,
+          createRoutesFromChildren,
+          matchRoutes,
+        }),
```
Comment (LetItRock) [inline apps/dashboard/src/utils/sentry.ts:19]: [react-router integration](https://docs.sentry.io/platforms/javascript/guides/react/features/react-router/)

File: apps/dashboard/src/utils/sentry.ts [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+        Sentry.replayIntegration({
+          maskAllText: true,
+          blockAllMedia: true,
+        }),
```
Comment (LetItRock) [inline apps/dashboard/src/utils/sentry.ts:23]: [replay integration](https://docs.sentry.io/platforms/javascript/guides/react/session-replay/)

File: apps/dashboard/src/utils/sentry.ts [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+        Sentry.captureConsoleIntegration({
+          levels: ['error'],
+        }),
```
Comment (LetItRock) [inline apps/dashboard/src/utils/sentry.ts:26]: [console logs capturing](https://docs.sentry.io/platforms/javascript/guides/react/configuration/integrations/captureconsole/)

File: apps/dashboard/src/utils/sentry.ts [hunk 14]
```diff
@@ -0,0 +1,64 @@
...
+        Sentry.browserTracingIntegration(),
```
Comment (LetItRock) [inline apps/dashboard/src/utils/sentry.ts:27]: [tracing](https://docs.sentry.io/platforms/javascript/guides/react/configuration/integrations/browsertracing/)

File: apps/dashboard/src/utils/sentry.ts [hunk 15]
```diff
@@ -0,0 +1,64 @@
...
+        Sentry.browserProfilingIntegration(),
```
Comment (LetItRock) [inline apps/dashboard/src/utils/sentry.ts:28]: [profiling](https://docs.sentry.io/platforms/javascript/guides/react/configuration/integrations/browserprofiling/)

File: apps/dashboard/src/utils/sentry.ts [hunk 16]
```diff
@@ -0,0 +1,64 @@
...
+       * sessions when an error occurs.
+       */
+      replaysOnErrorSampleRate: 1.0,
+      /*
+       * Set tracesSampleRate to 1.0 to capture 100%
+       * of transactions for performance monitoring.
+       * We recommend adjusting this value in production
+       */
+      tracesSampleRate: 1.0,
+      tracePropagationTargets: ['localhost', /^https:\/\/api\.novu\.co/, /^https:\/\/api\.novu-staging\.co/],
```
Comment (LetItRock) [inline apps/dashboard/src/utils/sentry.ts:55]: propagate traces to api

File: apps/dashboard/src/utils/types.ts [hunk 17]
```diff
@@ -19,5 +19,24 @@ export enum ConnectionStatus {
...
+  MISSING_VALUE = 'MISSING_VALUE',
+  WORKFLOW_ID_ALREADY_EXIST = 'WORKFLOW_ID_ALREADY_EXIST',
+  STEP_ID_ALREADY_EXIST = 'STEP_ID_ALREADY_EXIST',
+}
+
+export type RuntimeIssue = {
+  issueType: WorkflowIssueTypeEnum;
+  variableName?: string;
+  message: string;
+};
```
Comment (LetItRock) [inline apps/dashboard/src/utils/types.ts:34]: we will remove it later when BE types will be available

File: apps/dashboard/vite.config.ts [hunk 18]
```diff
@@ -25,10 +26,23 @@ export default defineConfig({
         });
       },
     },
+    // Put the Sentry vite plugin after all other plugins
+    sentryVitePlugin({
+      org: process.env.SENTRY_ORG,
+      project: process.env.SENTRY_PROJECT,
+      // Auth tokens can be obtained from https://sentry.io/orgredirect/organizations/:orgslug/settings/auth-tokens/
+      authToken: process.env.SENTRY_AUTH_TOKEN,
+      reactComponentAnnotation: { enabled: true },
```
Comment (LetItRock) [inline apps/dashboard/vite.config.ts:35]: [react components names](https://docs.sentry.io/platforms/javascript/guides/react/features/component-names/)

File: packages/shared/src/types/http/headers.types.ts [hunk 19]
```diff
@@ -6,6 +6,7 @@ export enum HttpRequestHeaderKeysEnum {
   USER_AGENT = 'User-Agent',
   CONTENT_TYPE = 'Content-Type',
   SENTRY_TRACE = 'Sentry-Trace',
+  BAGGAGE = 'Baggage',
```
Comment (LetItRock) [inline packages/shared/src/types/http/headers.types.ts:9]: this is the header attached by sentry for tracing propagation, [ref](https://docs.sentry.io/platforms/javascript/guides/react/tracing/trace-propagation/#basic-example)

File: apps/dashboard/src/components/workflow-editor/workflow-editor-provider.tsx (hunk 7 above)
Comment (SokratisVidros) [reply apps/dashboard/src/components/workflow-editor/workflow-editor-provider.tsx:72]: Let's add a TODO here to fix this in the future.

File: apps/dashboard/src/context/clerk-provider.tsx [hunk 20]
```diff
@@ -23,6 +23,7 @@ const CLERK_LOCALIZATION = {
     action__signOutAll: 'Log out from all accounts',
     action__manageAccount: 'Settings',
   },
+  formFieldLabel__organizationSlug: 'URL friendly identifier',
```
Comment (SokratisVidros) [inline apps/dashboard/src/context/clerk-provider.tsx]: Oops, this should go away with a latest rebase.

File: apps/dashboard/src/routes/root.tsx [hunk 21]
```diff
@@ -1,24 +1,44 @@
...
+      fallback={({ error, eventId }) => (
+        <>
+          Sorry, but something went wrong. <br />
+          Our team has been notified and we are investigating.
```
Comment (SokratisVidros) [inline apps/dashboard/src/routes/root.tsx]: Nit picky: we can add a prompt to contact support providing the event Id.

**Review by SokratisVidros — APPROVED** [review]
Nice!

**Pull request merged**